fix: formatting

RunDevelopment opened the pull request:
I rewrote large parts of the parser and the whole stringify and formatting code.

Changes:
- The Markdown formatting is now handled by Prettier's MD formatter.
- Print width is now correctly enforced everywhere.
- The 2 extra spaces after a description line break have been removed. (E.g. when the desc of an `@param` couldn't fit into one line, all lines after the first one used to be intended with 2 spaces.)
- Paragraphs are no longer capitalized. Since Prettier's MD formatter doesn't do this, we now also don't. The only exception to this is the first character of any description, this one will be capitalized.

---

This fixes #25.
This fixes #26.
This fixes #32. (The current fix is not correct)

hosseinmd commented:
Prettier markdown formatter is not good idea. I reverted that on this comment https://github.com/hosseinmd/prettier-plugin-jsdoc/commit/097f07d05ff062a61199bf67ddb2e6c486310223
Because the result of that is not correct.

hosseinmd commented:
Please split your PR to a few PR for example a PR for clean code and a PR for breaking changes

hosseinmd commented:
Prettier mark down has multiple bug. I tested that in big project like ant-design or the lodash

RunDevelopment commented:
> Prettier mark down has multiple bug.

The reason I switched to Prettier MD is that the current MD formatted of this plugin also has bugs. So [both have bugs](https://github.com/prettier/prettier/issues?q=is%3Aopen+label%3Alang%3Amarkdown+label%3Atype%3Abug), that's to be expected from two complex pieces of software. But I trust the Prettier Markdown formatter a lot more. The Prettier Markdown formatter is used by thousands of people daily, so bugs can be reported and fixed.

Put simply, I cannot believe that a formatter used by thousands and under active development is consistently worse than the regex-based formatter of this plugin.

> I tested that in big project like ant-design or the lodash

I tested this PR on lodash just now. I was not able to find any MD formatting bugs when using Prettier's MD formatter. I manually checked a sample of about 80 files and none of them had MD formatting errors.

May I ask what bugs specifically moved to creating your own markdown formatted?

> Please split your PR into a few PR

Sure!

RunDevelopment commented:
> > Please split your PR into a few PR
> 
> Sure!

Actually, I take that back. I wanted to split the PR into multiple feature PRs but this will be _very_ difficult. Do have more specific suggestions on how to split it?

You suggested a clean code PR, but I don't think that I did any major refactoring. I may have fixed some typos on the go but I don't know if it's worth making this its own PR.

hosseinmd commented on tests/__snapshots__/files.test.js.snap:
```diff
@@ -190,18 +190,18 @@ class test {
+       * @param {string | (substring: string, ...args: any[]) => string} A_big_string_for_test
+       * A_big_string_for_test string containing the text to replace for every
+       * successful match of searchValue in this string.
        * @returns {StarkStringType & NativeString}
```
print width not correct

RunDevelopment replied:
No, the print width is correct. It only seems too short because the `@param` line is too long. I know that the old code simply extended the print width in that case but I considered this bug.

hosseinmd commented on tests/__snapshots__/files.test.js.snap:
```diff
@@ -215,12 +215,12 @@ export interface FetchCallbackResponseArray<T, V> {
    * @deprecated Resolve clear with condition in your fetch api this function
-   *   will be remove
+   * will be remove
    */
```
Don't remove space of multi lines second line.

hosseinmd replied:
We are added this guide from JavaScript Style Guide.
https://google.github.io/styleguide/jsguide.html#jsdoc-method-and-function-comments

RunDevelopment replied:
May I ask why this was added in the first place?

It's not like this plugin follows the Google JS style guide closely. The indentation of exactly 2 spaces is somewhat arbitrary as well.

I also what to point out that the indentation is semantically invalid Markdown. Consistently indented lines are used for code blocks in MD. This means that other tools like JSDoc and TS may correctly assume that those are code blocks.

That being said, it does look good. But I don't know if good looks justify us breaking other tools.

hosseinmd replied:
The Markdown don't have tag description JSDoc is not markdown just part of JSDoc description can be like Markdown.

I used Markdown prettier formatter just few commits ago, reverted that, because has bad result, I don't like mistake again, at least on this time we can't use Markdown

hosseinmd commented:
Excuse me about approve that is mistaken.

RunDevelopment commented:
@hosseinmd 

If it's fine with you, then I will make a series of small PRs over next few days. These PRs will slowly introduce all features of this PR. I think this may be better since we can talk about changes in more detail. 

The only downside is that it will most likely take multiple days. The reason why I said that it's "very difficult" to split this PR is that all these smaller PRs depend on each other, so I have to make them sequentially one after another, so it's slower for me. I don't know how into many PRs I will split this up but it will probably be around 10.

hosseinmd commented:
So I close this.

Thank you
